### ISSUE TITLE
[1.12-rc2] service publishing to an existing published port fails silently

### ISSUE THREAD

**Issue by jhorwit2** [docker/docker#23831]
I'm running docker for mac beta

```
$ docker version
Client:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   906eacd
 Built:        Fri Jun 17 20:35:33 2016
 OS/Arch:      darwin/amd64
 Experimental: true

Server:
 Version:      1.12.0-rc2
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   a7119de
 Built:        Fri Jun 17 22:09:20 2016
 OS/Arch:      linux/amd64
 Experimental: true
```

I can run 
```
docker service create --replicas 3 --name nginx --publish 8081:80 nginx:latest
```

and it works fine but if i run another service that binds to port 8081 like 
```
docker service create --replicas 3 --name nginx2 --publish 8081:80 nginx:latest
```
it fails silently and stays 0.
```
ID            NAME    REPLICAS  IMAGE         COMMAND
5n1bppxuvboq  nginx2  0/3       nginx:latest
b7k4b2p1opfz  nginx   3/3       nginx:latest
```

**Comment by jhorwit2** [comment]
duplicate of https://github.com/docker/swarmkit/issues/939

**Comment by vdemeester** [comment]
This is kinda the same as #23709, maybe we should add a warn or something :angel: 
/cc @thaJeztah @mrjana

**Comment by jhorwit2** [comment]
Ah missed that ticket in my search. Feel free to close this one out if you want to fix it under that one. 👍

**Comment by jhorwit2** [comment]
Oh, actually after looking at that ticket more that's a completely different expected outcome. My thinking is that typically published ports are going to be known services that you can use to route external facing load balancers or w/e at. If you use the rolling upgrade feature and your service never goes down that task will forever be in limbo, which I feel like would be pretty common in this case.

**Comment by thaJeztah** [comment]
@jhorwit2 what do you think the expected behavior should be? In your example, you're asking docker to deploy two services, both using the same public port. The first service succeeds, because the port is available, the *second* is basically in a queue until the port becomes available. I think that behavior is correct (although we would like to have something so that you could check *why* it's "stuck"; see https://github.com/docker/docker/issues/23706).

It's a different way of looking at it; scheduling a conflicting port doesn't *fail*, because the service API is "declarative", i.e. I *want* something to be scheduled, and Docker makes sure to "eventually" get there. In this case "eventually" is "when the port becomes available".

**Comment by jhorwit2** [comment]
My use case comes from the [key concepts](https://docs.docker.com/engine/swarm/key-concepts/) from the docs. If I am publishing a service so that it's external to the swarm then I'm essentially agreeing to a contract with some outside service (load balancer, etc) that my service will exist on that port. I'm struggling to see a use case where you would want to schedule a service to take over the port once the current thing bound on that port is done.  

Perhaps a good meet in the middle would be another step in the flow when you are trying to create a service with a published port that's already taken. 

```
docker create --replicas 3 --name --publish 80:80 nginx nginx:latest
docker create --replicas 1 --name foobar --publish 80:80 hello_world:latest
> A published port is already published by another service. Do you want to schedule this service for when that port becomes available? [yN]
```
It seems the scheduler + service might be getting too complex. A nice thing about docker is it's very clear what you are running and what it does. When you start thinking about docker as a distributed scheduler with services that will "eventually" get scheduled that goes away. Perhaps create/update/scale should be blocking like terraform? That way you can provide feedback throughout the process which gets rid of any magic. The only way it won't be blocking is if you explicitly set it so as I feel the common use case is people want their thing deployed now, not later.

**Comment by thaJeztah** [comment]
> I'm struggling to see a use case where you would want to schedule a service to take over the port once the current thing bound on that port is done.

Possibly a new service, replacing an old service

It's not just at deploy time, but (for example), `docker service scale foo=300`, or `docker service update --image=hello_world:v2` are non-blocking as well; you express the *desired* state. I don't think asking confirmation is the right approach there, but being able to monitor what's happening (as mentioned earlier; https://github.com/docker/docker/issues/23706) sure is something that needs to be added.

**Comment by jhorwit2** [comment]
I thought about the replacement thing too, but why wouldn't you use update? If you do it via another create call then you will have downtime.

**Comment by jhorwit2** [comment]
You are expressing a desired state, but when you are dealing with applications you typically are going to say "scale to 300" then come back 3 hours later to see if your issue got solved. Right now you will have to keep inspecting your service to see that it got scaled. This means your release/update/scale doesn't provide you with feedback which seems a little strange for a release tool. Compare this to terraform, ansible, puppet, etc. From an operations perspective, I can't see where the async functionality would be ideal when monitoring a deployment. 

I could see async being useful when you are scheduling a task that you don't care about when it gets run, which is a not common use case I'd imagine for docker currently. As such making it a separate flag to be async seems pretty valid.

**Comment by tiborvass** [comment]
@thaJeztah do we have an issue tracking the async UX / error-reporting ? That way we can close this one since it's mixing conversation about a bug and UX.

**Comment by tiborvass** [comment]
Found it https://github.com/docker/docker/issues/23706. Closing if that's alright. Comment if you disagree we can reopen if needed.
